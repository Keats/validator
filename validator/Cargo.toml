[package]
name = "validator"
version = "0.16.0"
authors = ["Vincent Prouillet <hello@vincentprouillet.com"]
license = "MIT"
description = "Common validation functions (email, url, length, ...) and trait - to be used with `validator_derive`"
homepage = "https://github.com/Keats/validator"
repository = "https://github.com/Keats/validator"
keywords = ["validation", "api", "validator"]
edition = "2018"
readme = "../README.md"

[dependencies]
url = "2"
regex = "1"
lazy_static = "1"
idna = "0.2"
serde = "1"
serde_derive = "1"
serde_json = "1"
validator_derive = { version = "0.16", path = "../validator_derive", optional = true }
card-validate = { version = "2.2", optional = true }
phonenumber = { version = "0.3", optional = true }
unic-ucd-common = { version = "0.9", optional = true }
indexmap = {version = "1", features = ["serde-1"], optional = true }


[features]
phone = ["phonenumber", "validator_derive/phone"]
card = ["card-validate", "validator_derive/card"]
unic = ["unic-ucd-common", "validator_derive/unic"]
derive = ["validator_derive"]
