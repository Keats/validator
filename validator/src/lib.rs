//! # Example:
//!
//! ```ignore, no_run
//! use serde::Deserialize;
//!
//! // A trait that the Validate derive will impl
//! use validator::{Validate, ValidationError};
//!
//! #[derive(Debug, Validate, Deserialize)]
//! struct SignupData {
//!     #[validate(email)]
//!     mail: String,
//!     #[validate(phone)]
//!     phone: String,
//!     #[validate(url)]
//!     site: String,
//!     #[validate(length(min = 1), custom = "validate_unique_username")]
//!     #[serde(rename = "firstName")]
//!     first_name: String,
//!     #[validate(range(min = 18, max = 20))]
//!     age: u32,
//! }
//!
//! fn validate_unique_username(username: &str) -> Result<(), ValidationError> {
//!     if username == "xXxShad0wxXx" {
//!         // the value of the username will automatically be added later
//!         return Err(ValidationError::new("terrible_username"));
//!     }
//!
//!     Ok(())
//! }
//!
//! match signup_data.validate() {
//!   Ok(_) => (),
//!   Err(e) => return e;
//! };
//! ```
//!
//! # Available Validations:
//! | Validation              | Notes                                                 |
//! | ----------------------- | ----------------------------------------------------- |
//! | `email`                 |                                                       |
//! | `url`                   |                                                       |
//! | `length`                |                                                       |
//! | `range`                 |                                                       |
//! | `must_match`            |                                                       |
//! | `contains`              |                                                       |
//! | `custom`                |                                                       |
//! | `regex`                 |                                                       |
//! | `credit_card`           | (Requires the feature `card` to be enabled)           |
//! | `phone`                 | (Requires the feature `phone` to be enabled)          |
//! | `non_control_character` | (Required the feature `unic` to be enabled)           |
//! | `nested`                | (Uses the validation of the field type it self)       |
//! | `required`              |                                                       |
//!
//! [Checkout the project README of an in-depth usage description with examples.](https://github.com/Keats/validator/blob/master/README.md)
//!
//! # Installation:
//! Add the validator to the dependencies in the Cargo.toml file.
//!
//! ```toml
//! [dependencies]
//! validator = { version = "0.12", features = ["derive"] }
//! ```

mod traits;
mod types;
mod validation;

#[cfg(feature = "card")]
pub use validation::cards::validate_credit_card;
pub use validation::contains::validate_contains;
pub use validation::email::validate_email;
pub use validation::ip::{validate_ip, validate_ip_v4, validate_ip_v6};
pub use validation::length::validate_length;
pub use validation::must_match::validate_must_match;
#[cfg(feature = "unic")]
pub use validation::non_control_character::validate_non_control_character;
#[cfg(feature = "phone")]
pub use validation::phone::validate_phone;
pub use validation::range::validate_range;

pub use validation::required::validate_required;
pub use validation::urls::validate_url;
pub use validation::Validator;

pub use traits::{Contains, HasLen, Validate, ValidateArgs};
pub use types::{ValidationError, ValidationErrors, ValidationErrorsKind};

#[cfg(feature = "derive")]
pub use validator_derive::Validate;
