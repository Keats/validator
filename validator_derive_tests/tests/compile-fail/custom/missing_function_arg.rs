use validator::Validate;

#[derive(Validate)]
struct Test {
    #[validate(custom(arg = "i64"))]
    s: String,
}

fn main() {}
