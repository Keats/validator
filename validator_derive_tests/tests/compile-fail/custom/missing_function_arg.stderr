error: Invalid attribute #[validate] on field `s`: Validator `custom` requires the function parameter
 --> $DIR/missing_function_arg.rs:5:5
  |
5 |     #[validate(custom(arg = "i64"))]
  |     ^
